ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java

```diff
@@ -398,7 +398,24 @@ private boolean deleteIfNeeded(BlobId blobId, long deletionTime) throws CloudSto
 
   @Override
   public short undelete(MessageInfo info) throws StoreException {
-    throw new UnsupportedOperationException("Undelete not supported in cloud store");
+    // TODO: Currently this is implemented for undeletes via replication only for DR.
+    checkStarted();
+    try {
+      BlobId blobId = (BlobId) info.getStoreKey();
+      // We do not have enough information in cache to check life version.
+      // It is expected that un-deletes will be very rare, so its ok to trade off occasional hits to azure,
+      // in lieu of maintaining live version along with blob id in cache.
+      short newLifeVersion =
+          requestAgent.doWithRetries(() -> cloudDestination.undeleteBlob(blobId, info.getLifeVersion()), "Undelete",
```

Comment:
I want to point out that info.getLifeVersion might return -1 (MessageInfo.LIFE_VERSION_FROM_FRONTEND). If so, it means we have to fetch the current lifeVersion and increment it and then put it back to the azure and cosmo. Can you add a todo here so later we can deal with it.

Reply:
done